### PR TITLE
Preserve same vtable pointer when cloning raw waker, to fix Waker::will_wake

### LINKED ISSUE #121600: Waker::will_wake() broken on nightly
I would expect this to run to completion:

```rust
use std::sync::Arc;
use std::task::{Wake, Waker};

struct NoopWaker;

impl Wake for NoopWaker {
    fn wake(self: Arc<Self>) {}
}

fn main() {
    let waker: Waker = Arc::new(NoopWaker).into();
    let other = waker.clone();

    assert!(waker.will_wake(&other));
}
```

On stable, it does:

```
$ rustup default stable
info: using existing install for 'stable-x86_64-apple-darwin'
info: default toolchain set to 'stable-x86_64-apple-darwin'

  stable-x86_64-apple-darwin unchanged - rustc 1.76.0 (07dca489a 2024-02-04)

$ cargo run
   Compiling will_wake v0.1.0 (/Users/justin/dev/will_wake)
    Finished dev [unoptimized + debuginfo] target(s) in 0.38s
     Running `target/debug/will_wake`
```

But not on current nightly:

```
$ rustup default nightly
info: using existing install for 'nightly-x86_64-apple-darwin'
info: default toolchain set to 'nightly-x86_64-apple-darwin'

  nightly-x86_64-apple-darwin unchanged - rustc 1.78.0-nightly (381d69953 2024-02-24)

$ cargo run
   Compiling will_wake v0.1.0 (/Users/justin/dev/will_wake)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.29s
     Running `target/debug/will_wake`
thread 'main' panicked at src/main.rs:14:5:
assertion failed: waker.will_wake(&other)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Interestingly, it works ok without cargo:

```
$ rustc --edition 2021 src/main.rs
$ ./main
```

The cargo project isn't special though:

```
$ cat Cargo.toml 
[package]
name = "will_wake"
version = "0.1.0"
edition = "2021"

[dependencies]
```

OS: macOS 13.6.3

### PR DESCRIPTION
**Pull request by dtolnay**
Fixes #121600.

As @jkarneges identified in https://github.com/rust-lang/rust/issues/121600#issuecomment-1963041051, the issue is two different const promotions produce two statics at different addresses, which may or may not later be deduplicated by the linker (in this case not).

Prior to #119863, the content of the statics was compared, and they were equal. After, the address of the statics are compared and they are not equal.

It is documented that `will_wake` _"works on a best-effort basis, and may return false even when the Wakers would awaken the same task"_ so this PR fixes a quality-of-implementation issue, not a correctness issue.

### CODE REVIEW

**Comment by dtolnay** [comment]
@tmiasko good call, I confirmed that works.

**Comment by dtolnay** [comment]
I am uncertain whether we are better off with this change (always inline clone_waker) or with reverting #119863.

Are there cases where `Waker::will_wake` will return true prior to #119863, and will return false after this PR? In other words, cases where the program would end up with multiple identical copies of `RawWakerVTable::new(..., wake::<W>, ...)` all referring to the same instantiation of `wake::<W>` etc.

If the deduplication of structurally equivalent constants is done by rustc at LLVM IR code generation time while the deduplication of generic function instantiations is done at link time then potentially yes, right?

**Comment by dtolnay** [comment]
Thinking about it more, I think this PR is optimal, at least for the LLVM backend. There is no way that cloning a RawWaker returned by raw_waker() would produce a clone that has a different vtable pointer than the original.

Not sure about Miri, or any non-LLVM backend, but that's what the "best-effort basis" part of `will_wake` is for.

**Comment by cuviper** [comment]
Makes sense to me.

@bors r+

File: library/alloc/tests/task.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    impl Wake for NoopWaker {
+        fn wake(self: Arc<Self>) {}
+    }
+
+    let waker = Waker::from(Arc::new(NoopWaker));
+    let clone = waker.clone();
+
+    assert!(waker.will_wake(&clone));
+    assert!(clone.will_wake(&waker));
```
Comment (RalfJung) [inline library/alloc/tests/task.rs:18]: These tests are testing things which are not guaranteed by the language, they just happen to be what rustc currently does. In particular, these tests [fail in Miri](https://github.com/rust-lang/miri-test-libstd/actions/runs/8135079336/job/22229021537).

Not sure if that is intentional; there's no comment here indicating such an intent.

**Comment by taiki-e** [comment]
I tested this in https://github.com/rust-lang/futures-rs/pull/2834, but this seems to only work in debug build.
In release mode, a test that passed before https://github.com/rust-lang/rust/pull/119863 now fails. https://github.com/rust-lang/futures-rs/actions/runs/8129901894/job/22233663120?pr=2834

File: library/alloc/tests/task.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+#[test]
+fn test_local_waker_will_wake_clone() {
+    struct NoopWaker;
+
+    impl LocalWake for NoopWaker {
+        fn wake(self: Rc<Self>) {}
+    }
+
+    let waker = LocalWaker::from(Rc::new(NoopWaker));
```
Comment (RalfJung) [inline library/alloc/tests/task.rs:29]: More problems with this test... it seems to have a memory leak. Is that expected?
```
0.107460   error: memory leaked: alloc3172879 (Rust heap, size: 16, align: 8), allocated here:
  0.000050      --> /home/runner/work/miri-test-libstd/miri-test-libstd/rust-src-patched/library/alloc/src/alloc.rs:100:9
  0.000009       |
  0.000007   100 |         __rust_alloc(layout.size(), layout.align())
  0.000006       |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  0.000006       |
  0.000005       = note: BACKTRACE:
  0.000006       = note: inside `std::alloc::alloc` at /home/runner/work/miri-test-libstd/miri-test-libstd/rust-src-patched/library/alloc/src/alloc.rs:100:9: 100:52
  0.000005       = note: inside `std::alloc::Global::alloc_impl` at /home/runner/work/miri-test-libstd/miri-test-libstd/rust-src-patched/library/alloc/src/alloc.rs:183:73: 183:86
  0.000007       = note: inside `<std::alloc::Global as std::alloc::Allocator>::allocate` at /home/runner/work/miri-test-libstd/miri-test-libstd/rust-src-patched/library/alloc/src/alloc.rs:243:9: 243:39
  0.000011       = note: inside `alloc::alloc::exchange_malloc` at /home/runner/work/miri-test-libstd/miri-test-libstd/rust-src-patched/library/alloc/src/alloc.rs:332:11: 332:34
  0.000011       = note: inside `std::boxed::Box::<std::rc::RcBox<task::test_local_waker_will_wake_clone::NoopWaker>>::new` at /home/runner/work/miri-test-libstd/miri-test-libstd/rust-src-patched/library/alloc/src/boxed.rs:218:9: 218:20
  0.000010       = note: inside `std::rc::Rc::<task::test_local_waker_will_wake_clone::NoopWaker>::new` at /home/runner/work/miri-test-libstd/miri-test-libstd/rust-src-patched/library/alloc/src/rc.rs:398:27: 398:94
  0.000011   note: inside `task::test_local_waker_will_wake_clone`
  0.000010      --> alloc_miri_test/../library/alloc/tests/task.rs:31:34
  0.000009       |
  0.000009   31  |     let waker = LocalWaker::from(Rc::new(NoopWaker));
  0.000009       |                                  ^^^^^^^^^^^^^^^^^^
```
Any idea how to fix that?

Comment (RalfJung) [reply library/alloc/tests/task.rs:29]: Filed as https://github.com/rust-lang/rust/issues/122180

**Pull request merged**